benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java

```diff
@@ -150,52 +158,145 @@ private BenchmarkExecutionResult failedBenchmarkResult(Benchmark benchmark, Exce
-                        }
-                        QueryExecutionResultBuilder failureResult = new QueryExecutionResultBuilder(queryExecution)
-                                .startTimer();
-                        try {
-                            result = queryExecutionDriver.execute(queryExecution, connection);
-                        }
-                        catch (Exception e) {
-                            LOG.error("Query Execution failed for benchmark {}", benchmark.getName());
-                            result = failureResult
-                                    .endTimer()
-                                    .failed(e)
-                                    .build();
-                        }
+    private List<Callable<List<QueryExecutionResult>>> buildConcurrencyQueryExecutionCallables(Benchmark benchmark, int runs, boolean reportStatus, Optional<ZonedDateTime> lastTime)
+    {
+        List<Callable<List<QueryExecutionResult>>> executionCallables = newArrayList();
+        for (int thread = 0; thread < benchmark.getConcurrency(); thread++) {
+            int finalThread = thread;
+            executionCallables.add(() -> {
+                LOG.info("Running throughput test: {} queries, {} runs", benchmark.getQueries().size(), runs);
+                Integer[] queryOrder = PermutationUtils.preparePermutation(benchmark.getQueries().size(), finalThread);
+                List<QueryExecutionResult> queryExecutionResults = executeConcurrentQueries(benchmark, runs, reportStatus, lastTime, finalThread, queryOrder);
+                if (reportStatus && !queryExecutionResults.isEmpty()) {
+                    statusReporter.reportConcurrencyTestExecutionFinished(queryExecutionResults);
+                }
+                return queryExecutionResults;
+            });
+        }
+        return executionCallables;
+    }
 
-                        if (reportStatus) {
-                            statusReporter.reportExecutionFinished(result);
+    private List<QueryExecutionResult> executeConcurrentQueries(Benchmark benchmark, int runs, boolean reportStatus, Optional<ZonedDateTime> lastTime, int threadNumber, Integer[] queryOrder)
+            throws SQLException
+    {
+        boolean firstQuery = true;
+        List<QueryExecutionResult> queryExecutionResults = newArrayList();
+        try (Connection connection = getConnectionFor(new QueryExecution(benchmark, benchmark.getQueries().get(0), 0))) {
+            for (int run = 1; run <= runs; run++) {
+                for (int queryIndex = 0; queryIndex < benchmark.getQueries().size(); queryIndex++) {
+                    int permutedQueryIndex = queryIndex;
+                    if (!reportStatus) {
+                        LOG.info("Comparing {} vs {}, {}, {}", queryIndex % benchmark.getConcurrency(), threadNumber, queryIndex, benchmark.getConcurrency());
+                        if (queryIndex % benchmark.getConcurrency() != threadNumber) {
```

Comment:
does pre-warm account for total query time? Is pre-warm a separate stage before actual run?

Reply:
Pre-warm is a separate stage, results of pre-warm runs are not reported - see https://github.com/ppalucha/benchto/blob/pp/concurrency-round-robin-test/benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java#L107

For "standard" benchmark usually each benchmark contains only single query, therefore pre-warming runs for specific query are run just before this query. For concurrency tests, benchmark contains multiple queries. Pre-warming runs for all queries are executed before measured executions.

Reply:
OK, but are all threads waiting for pre-warm to complete before running actual queries?

Reply:
Yes, they wait.